### ISSUE TITLE
`save_last` should only keep the most recent checkpoint (along with the top k)

### ISSUE THREAD

**Issue by ZhaofengWu** [PyTorchLightning/pytorch-lightning#2141]
## 🚀 Feature
After `save_last` saves a checkpoint, it removes the previous "last" (i.e. latest) checkpoint (i.e. separate from top k).

### Motivation

For example, for someone limited by disk space, a good strategy during training would be to always save the best checkpoint as well as the latest checkpoint to restore from in case training gets interrupted (and ideally with an option to automatically remove the latest checkpoint with training finishes, but that might belong to a separate issue). Setting `save_top_k=1` and `save_last=True` almost achieves this, but the latter actually results in all checkpoints being saved, which is unnecessary for the purpose of restoring training. So I propose `save_last` to keep only one latest checkpoint at all times. This is separate from the top k checkpoints. If a user really wants to keep all checkpoints around, he/she can always do `save_top_k=-1` anyway.

### Alternatives

Some other config option that allows always keeping one latest checkpoint.

**Comment by carmocca** [comment]
I don't understand this issue

> After save_last saves a checkpoint, it removes the previous "last" checkpoint (i.e. separate from top k).

Doesn't it work like this already?

**Comment by ZhaofengWu** [comment]
If there have been no changes in the flag's behavior in the past few months, I think setting `save_last` always results in all checkpoints being saved.

**Comment by carmocca** [comment]
You can have a look at the code to save last here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/a2b961c7bc0d859814f167e471fbc24e4fde7b1b/pytorch_lightning/callbacks/model_checkpoint.py#L467-L487

Every epoch end, the "last" model is saved and the one from the previous epoch is deleted. So only one is kept when training is finished.

**Comment by ZhaofengWu** [comment]
Looking at the blame there have been a lot of changes since I opened this issue. Perhaps this is fixed now. I haven't confirmed this, but feel free to close this issue if the behavior is correct now.

**Comment by awaelchli** [comment]
Yep, there have been some fixes and improvements. We've also added more tests around topk and save last so that both can be used. I believe what you describe here now correctly works. In summary:

```
save_last=True:            saves a last.ckpt
save_top_k=1:              saves the top checkpoint (you will have at most 1 file)
save_top_k=n:              you have n files saved with epoch numbers
save_top_k=n, save_last:   you have n files + "last.ckpt" 
save_top_k=-1:             saves all
```

and the default model checkpoint Trainer(checkpoint_callback=True) saves every epoch the latest checkpoint, but we keep only one file around, this should be a good default for most users while saving space.

**Comment by hiyyg** [comment]
Hey @awaelchli , I wonder if there is a way to save last k checkpoints instead of just the latest checkpoint?

**Comment by carmocca** [comment]
@hiyyg there isn't anything like `save_last=5` but you could work around this by setting `save_top_k=5` and monitoring a metric that always improves, `monitor="global_step"` for example

**Comment by hiyyg** [comment]
Thanks @carmocca! I guess this is not what I want. `save_last=5` shoud save the last 5 checkpoints with some pre-defined checkpointing intervals, rather than the last 5 `global_step` checkpoints.

**Comment by carmocca** [comment]
That could still work, you can do (roughly)

```python
def training_step():
    ...
    if on_predefined_checkpoint_interval():
        self.log("checkpoint_signal", self.global_step, reduce_fx="max")
    ...

ModelCheckpoint(save_top_k=5, monitor="checkpoint_signal")
```

**Comment by dljjqy** [comment]
The one who design CHECKPOINT Module are really stupid.It is too useless than just use pytorch only.The only use of lightning is telling every coder they need to write a trainer by theri self but not use the lightning diriectly.Just use torch.save for god.

**Comment by ptn24** [comment]
I think the Lightning 2.2.1 documentation can be more clear on save_last. It says

> When True, saves a last.ckpt copy whenever a checkpoint file gets saved.

but the comments in this issue seem to say that the last.ckpt checkpoint gets saved every epoch, separately from the save_top_k checkpoints and whatnot.

**Comment by DrewGalbraith** [comment]
I'd like to 2nd this. I came here bc [latest docs](https://lightning.ai/docs/pytorch/1.6.0/common/checkpointing.html#:~:text=You%20can%20save%20the%20last,the%20monitor%20and%20save_top_k%20argument.) don't explain how these two setting interact, but @awaelchli's [comment](https://github.com/Lightning-AI/pytorch-lightning/issues/2141#issuecomment-700981674) lays it out perfectly. Can we get docs updated?
